File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/BurnerPairClient.kt
package com.flagshipserver.app.core

import kotlinx.coroutines.channels.Channel
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.net.URLEncoder

/**
 * Phone-side peer of `wss://<host>/burner-pipe/<sid>?role=phone` — the
 * one-shot deposit session with the desktop burner (relay DO:
 * apps/com/src/burnerRelay.ts). Kotlin mirror of the iOS BurnerPairClient.
 */

sealed interface BurnerInbound {
    data object Accepted : BurnerInbound
    data object PeerPresent : BurnerInbound
    data object PeerJoined : BurnerInbound
    data class BurnerHello(val burnerPkB64: String) : BurnerInbound
    data object RecipeAccepted : BurnerInbound
    // Phase 4 — the burner asks the phone to approve a security-sensitive
    // setting (e.g. Debug mode). Carries the box's serverDomain so the phone
    // can sign the owner-IRK debug-access grant the box verifies. Mirrors the
    // iOS BurnerInbound.consentRequest(setting, serverDomain, warning).
    data class ConsentRequest(val setting: String, val serverDomain: String, val warning: String) : BurnerInbound
    data object PeerGone : BurnerInbound
    data object Expired : BurnerInbound
    data class RelayError(val message: String) : BurnerInbound
    data object Pong : BurnerInbound
}

sealed interface BurnerOutbound {
    data class PhoneHello(val phonePkB64: String) : BurnerOutbound
    data object ConfirmPairing : BurnerOutbound
    data class Deliver(val ciphertextB64: String, val nonceB64: String) : BurnerOutbound
    data class Raw(val json: String) : BurnerOutbound

    fun toJson(): String = when (this) {
        is PhoneHello -> """{"kind":"phone-hello","phonePk":"$phonePkB64"}"""
        is ConfirmPairing -> """{"kind":"confirm-pairing"}"""
        is Deliver -> """{"kind":"deliver","ciphertext":"$ciphertextB64","nonce":"$nonceB64"}"""
        is Raw -> json
    }
}

interface BurnerPairClient {
    /** Open the WS as role=phone; returns a channel of inbound events. */
    suspend fun connect(sid: String): Channel<BurnerInbound>
    suspend fun send(frame: BurnerOutbound)
    fun close()
}

class LiveBurnerPairClient(
    private val client: OkHttpClient = OkHttpClient(),
    private val host: String = Endpoints.controlHost,
    secure: Boolean = true,
) : BurnerPairClient {
    private val scheme = if (secure) "wss" else "ws"
    private val json = Json { ignoreUnknownKeys = true }
    @Volatile private var ws: WebSocket? = null
    private var pinger: Thread? = null
    @Volatile private var closed = false
    @Volatile private var generation = 0
    private var channel = Channel<BurnerInbound>(Channel.UNLIMITED)

    override suspend fun connect(sid: String): Channel<BurnerInbound> {
        val previousWs = ws
        val previousChannel = channel
        pinger?.interrupt()
        generation += 1
        val connectionGeneration = generation
        closed = false
        val nextChannel = Channel<BurnerInbound>(Channel.UNLIMITED)
        channel = nextChannel
        val encodedSid = URLEncoder.encode(sid, "UTF-8")
        val req = Request.Builder()
            .url("$scheme://$host/burner-pipe/$encodedSid?role=phone")
            .build()
        val nextWs = client.newWebSocket(req, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                if (isCurrent(connectionGeneration, webSocket)) {
                    decode(text)?.let { nextChannel.trySend(it) }
                }
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (isCurrent(connectionGeneration, webSocket)) {
                    nextChannel.trySend(BurnerInbound.RelayError("connection lost"))
                    nextChannel.close()
                }
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                if (isCurrent(connectionGeneration, webSocket)) {
                    nextChannel.trySend(BurnerInbound.PeerGone)
                    nextChannel.close()
                }
            }
        })
        ws = nextWs
        previousWs?.cancel()
        previousChannel.close()
        startPing(connectionGeneration, nextWs)
        return nextChannel
    }

    override suspend fun send(frame: BurnerOutbound) {
        check(ws?.send(frame.toJson()) == true) { "the pairing connection is closed" }
    }

    private fun startPing(connectionGeneration: Int, socket: WebSocket) {
        pinger = Thread {
            while (!closed && generation == connectionGeneration) {
                try { Thread.sleep(20_000) } catch (_: InterruptedException) { return@Thread }
                if (closed || generation != connectionGeneration) return@Thread
                socket.send("""{"kind":"ping"}""")
            }
        }.apply { isDaemon = true; start() }
    }

    @Synchronized
    override fun close() {
        if (closed) return
        closed = true
        generation += 1
        pinger?.interrupt()
        ws?.close(1000, null)
        ws = null
        channel.close()
    }

    @Synchronized
    private fun isCurrent(connectionGeneration: Int, socket: WebSocket): Boolean =
        !closed && generation == connectionGeneration && ws === socket

    private fun decode(text: String): BurnerInbound? {
        val obj = runCatching { json.decodeFromString(JsonObject.serializer(), text) }.getOrNull() ?: return null
        val kind = obj["kind"]?.jsonPrimitive?.content ?: return null
        return when (kind) {
            "accepted" -> BurnerInbound.Accepted
            "peer-present" -> BurnerInbound.PeerPresent
            "peer-joined" -> BurnerInbound.PeerJoined
            "peer-gone" -> BurnerInbound.PeerGone
            "expired" -> BurnerInbound.Expired
            "pong" -> BurnerInbound.Pong
            "error" -> BurnerInbound.RelayError(obj["reason"]?.jsonPrimitive?.content ?: "relay error")
            "peer" -> {
                val frame = obj["frame"]?.jsonObject ?: return null
                when (frame["kind"]?.jsonPrimitive?.content) {
                    "burner-hello" -> frame["burnerPk"]?.jsonPrimitive?.content?.let { BurnerInbound.BurnerHello(it) }
                    "recipe-accepted" -> BurnerInbound.RecipeAccepted
                    "consent-request" -> BurnerInbound.ConsentRequest(
                        frame["setting"]?.jsonPrimitive?.content ?: "",
                        frame["serverDomain"]?.jsonPrimitive?.content ?: "",
                        frame["warning"]?.jsonPrimitive?.content ?: "",
                    )
                    else -> null
                }
            }
            else -> null
        }
    }
}

/** Scripted mock for tests. emit() pushes inbound; sent captures outbound. */
class MockBurnerPairClient : BurnerPairClient {
    var connectedSid: String? = null; private set
    var connectCount: Int = 0; private set
    val sent = mutableListOf<BurnerOutbound>()
    var didClose = false; private set
    private val channel = Channel<BurnerInbound>(Channel.UNLIMITED)

    override suspend fun connect(sid: String): Channel<BurnerInbound> {
        connectedSid = sid
        connectCount += 1
        return channel
    }
    override suspend fun send(frame: BurnerOutbound) { sent.add(frame) }
    override fun close() { didClose = true; channel.close() }

    fun emit(ev: BurnerInbound) { channel.trySend(ev) }
    val sentJson: List<String> get() = sent.map { it.toJson() }
}

File: apps/mobile/shared/Sources/FlagshipAPI/Client/BurnerPairClient.swift
import Foundation

/// Phone-side peer of `wss://<host>/burner-pipe/<sid>?role=phone` — the
/// bidirectional deposit session with the desktop burner (relay DO:
/// apps/com/src/burnerRelay.ts). The phone learns the burner's pubkey,
/// confirms the SAS, and delivers the recipe. A brief phone-side socket loss
/// reconnects to the same relay slot while pairing; once the burner returns a
/// successful staging receipt, the phone has no further role.

/// Decoded inbound events the phone cares about.
public enum BurnerInbound: Sendable, Equatable {
    case accepted
    case peerPresent           // the burner was already connected when we joined
    case peerJoined            // the burner joined after us
    case burnerHello(burnerPkB64: String)
    case recipeAccepted        // the burner successfully staged the recipe
    case consentRequest(setting: String, serverDomain: String, warning: String)
    case peerGone
    case expired
    case relayError(String)
    case pong
}

/// Frames the phone sends to the burner (forwarded verbatim by the relay).
public enum BurnerOutbound: Sendable {
    case phoneHello(phonePkB64: String)
    case confirmPairing
    case deliver(ciphertextB64: String, nonceB64: String)
    case raw(json: String)     // consent-result etc.
}

public protocol BurnerPairClient: Sendable {
    /// Open the WS as role=phone and return a stream of inbound events.
    func connect(sid: String) async throws -> AsyncStream<BurnerInbound>
    func send(_ frame: BurnerOutbound) async throws
    func close() async
}

public enum BurnerPairError: Error, LocalizedError, Sendable {
    case connectionFailed(String)
    public var errorDescription: String? {
        switch self {
        case .connectionFailed(let m): return "Couldn't reach the relay: \(m)"
        }
    }
}

extension BurnerOutbound {
    var json: String {
        switch self {
        case .phoneHello(let pk):
            return Self.obj(["kind": "phone-hello", "phonePk": pk])
        case .confirmPairing:
            return Self.obj(["kind": "confirm-pairing"])
        case .deliver(let ct, let nonce):
            return Self.obj(["kind": "deliver", "ciphertext": ct, "nonce": nonce])
        case .raw(let json):
            return json
        }
    }
    private static func obj(_ d: [String: String]) -> String {
        (try? JSONSerialization.data(withJSONObject: d)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    }
}

// MARK: - Live

public final class LiveBurnerPairClient: BurnerPairClient, @unchecked Sendable {
    public static var defaultHost: String { Endpoints.controlHost }

    private let urlSession: URLSession
    private let host: String
    private let scheme: String
    private let stateLock = NSLock()
    private var task: URLSessionWebSocketTask?
    private var continuation: AsyncStream<BurnerInbound>.Continuation?
    private var connectionID: UUID?

    public init(urlSession: URLSession = .shared, host: String = defaultHost, secure: Bool = true) {
        self.urlSession = urlSession
        self.host = host
        self.scheme = secure ? "wss" : "ws"
    }

    public func connect(sid: String) async throws -> AsyncStream<BurnerInbound> {
        let encodedSid = sid.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? sid
        guard let url = URL(string: "\(scheme)://\(host)/burner-pipe/\(encodedSid)?role=phone") else {
            throw BurnerPairError.connectionFailed("bad URL")
        }
        let t = urlSession.webSocketTask(with: url)
        var streamContinuation: AsyncStream<BurnerInbound>.Continuation!
        let stream = AsyncStream<BurnerInbound> { cont in
            streamContinuation = cont
        }
        let id = UUID()
        let previous = replaceConnection(task: t, continuation: streamContinuation, id: id)
        previous.continuation?.finish()
        previous.task?.cancel(with: .goingAway, reason: nil)
        t.resume()
        receiveLoop(task: t, id: id)
        startPing(task: t, id: id)
        return stream
    }

    public func send(_ frame: BurnerOutbound) async throws {
        guard let t = currentTask() else {
            throw BurnerPairError.connectionFailed("the connection is closed")
        }
        do {
            try await t.send(.string(frame.json))
        } catch {
            throw BurnerPairError.connectionFailed(Self.describe(error))
        }
    }

    public func close() async {
        let previous = clearConnection()
        previous.continuation?.finish()
        previous.task?.cancel(with: .normalClosure, reason: nil)
    }

    private func receiveLoop(task: URLSessionWebSocketTask, id: UUID) {
        task.receive { [weak self] result in
            guard let self, self.isCurrent(id) else { return }
            switch result {
            case .failure(let error):
                self.finishConnection(id: id, event: .relayError(Self.describe(error)))
            case .success(let message):
                if let ev = Self.decode(message) { self.yield(ev, id: id) }
                if self.isCurrent(id) { self.receiveLoop(task: task, id: id) }
            }
        }
    }

    private func startPing(task: URLSessionWebSocketTask, id: UUID) {
        Task { [weak self] in
            while true {
                try? await Task.sleep(nanoseconds: 20_000_000_000)
                guard let self, self.isCurrent(id) else { return }
                do {
                    try await task.send(.string("{\"kind\":\"ping\"}"))
                } catch {
                    self.finishConnection(id: id, event: .relayError(Self.describe(error)))
                    return
                }
            }
        }
    }

    private typealias Connection = (
        task: URLSessionWebSocketTask?,
        continuation: AsyncStream<BurnerInbound>.Continuation?
    )

    private func replaceConnection(
        task newTask: URLSessionWebSocketTask,
        continuation newContinuation: AsyncStream<BurnerInbound>.Continuation,
        id: UUID
    ) -> Connection {
        stateLock.lock()
        defer { stateLock.unlock() }
        let previous = (task, continuation)
        task = newTask
        continuation = newContinuation
        connectionID = id
        return previous
    }

    private func clearConnection() -> Connection {
        stateLock.lock()
        defer { stateLock.unlock() }
        let previous = (task, continuation)
        task = nil
        continuation = nil
        connectionID = nil
        return previous
    }

    private func currentTask() -> URLSessionWebSocketTask? {
        stateLock.lock()
        defer { stateLock.unlock() }
        return task
    }

    private func isCurrent(_ id: UUID) -> Bool {
        stateLock.lock()
        defer { stateLock.unlock() }
        return connectionID == id
    }

    private func yield(_ ev: BurnerInbound, id: UUID) {
        stateLock.lock()
        let current = connectionID == id ? continuation : nil
        stateLock.unlock()
        current?.yield(ev)
    }

    private func finishConnection(id: UUID, event: BurnerInbound) {
        stateLock.lock()
        guard connectionID == id else {
            stateLock.unlock()
            return
        }
        let current = continuation
        task = nil
        continuation = nil
        connectionID = nil
        stateLock.unlock()
        current?.yield(event)
        current?.finish()
    }

    private static func describe(_ error: Error) -> String {
        let ns = error as NSError
        return "Connection lost (\(ns.domain) \(ns.code)): \(ns.localizedDescription)"
    }

    private static func decode(_ message: URLSessionWebSocketTask.Message) -> BurnerInbound? {
        let data: Data
        switch message {
        case .data(let d): data = d
        case .string(let s): data = Data(s.utf8)
        @unknown default: return nil
        }
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let kind = obj["kind"] as? String else { return nil }
        switch kind {
        case "accepted": return .accepted
        case "peer-present": return .peerPresent
        case "peer-joined": return .peerJoined
        case "peer-gone": return .peerGone
        case "expired": return .expired
        case "pong": return .pong
        case "error": return .relayError((obj["reason"] as? String) ?? "relay error")
        case "peer":
            guard let frame = obj["frame"] as? [String: Any],
                  let fkind = frame["kind"] as? String else { return nil }
            switch fkind {
            case "burner-hello":
                guard let pk = frame["burnerPk"] as? String else { return nil }
                return .burnerHello(burnerPkB64: pk)
            case "recipe-accepted":
                return .recipeAccepted
            case "consent-request":
                let setting = (frame["setting"] as? String) ?? ""
                let serverDomain = (frame["serverDomain"] as? String) ?? ""
                let warning = (frame["warning"] as? String) ?? ""
                return .consentRequest(setting: setting, serverDomain: serverDomain, warning: warning)
            default:
                return nil
            }
        default:
            return nil
        }
    }
}

// MARK: - Mock

/// Scripted mock for tests. Push inbound events with `emit(_:)`; outbound
/// frames are captured in `sent`.
public final class MockBurnerPairClient: BurnerPairClient, @unchecked Sendable {
    public private(set) var connectedSid: String?
    public private(set) var connectCount = 0
    public private(set) var sent: [BurnerOutbound] = []
    public private(set) var didClose = false
    private var continuation: AsyncStream<BurnerInbound>.Continuation?

    public init() {}

    public func connect(sid: String) async throws -> AsyncStream<BurnerInbound> {
        connectedSid = sid
        connectCount += 1
        return AsyncStream { cont in self.continuation = cont }
    }

    public func send(_ frame: BurnerOutbound) async throws { sent.append(frame) }

    public func close() async { didClose = true; continuation?.finish(); continuation = nil }

    /// Test helper — emit an inbound event to the consumer.
    public func emit(_ ev: BurnerInbound) { continuation?.yield(ev) }

    /// Convenience accessors for assertions.
    public var sentJSON: [String] { sent.map { $0.json } }
}
